test/unit/test_chain_weight_clamp.cpp

```
@@ -0,0 +1,104 @@
+    // MEM_CHAIN_W_MAX + 1 -- the smallest value the old (1<<30)-1 clamp let
+    // through to wrap, and it wrapped to 0.
+    const int over = weight_of(seeds, 2, static_cast<int64_t>(1) << 26);
+    CHECK(over == static_cast<int>(MEM_CHAIN_W_MAX));
+
+    // Far over: the old clamp returned (1<<30)-1 here, which stores as
+    // (1<<30)-1 mod 2^27 == MEM_CHAIN_W_MAX -- accidentally right at the clamp
+    // ceiling but wrong for everything below it. Pin the saturation directly.
+    const int way_over = weight_of(seeds, 8, static_cast<int64_t>(1) << 26);
+    CHECK(way_over == static_cast<int>(MEM_CHAIN_W_MAX));
+}
+
```

Cppcheck (2.21.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)